src/librustc_mir/interpret/eval_context.rs

```diff
@@ -1409,47 +1439,223 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
+                    Place::from_ptr(ptr, ptr_align),
+                    layout,
+                )?;
+                variant = variant_index;
+                layout = layout.for_variant(self, variant_index);
+                trace!("variant layout: {:#?}", layout);
+            },
+            layout::Variants::Single { index } => variant = index,
+        }
+        match layout.fields {
+            // primitives are unions with zero fields
+            layout::FieldPlacement::Union(0) => {
+                match layout.abi {
+                    // nothing to do, whatever the pointer points to, it is never going to be read
+                    layout::Abi::Uninhabited => validation_failure!("a value of an uninhabited type", path),
+                    // check that the scalar is a valid pointer or that its bit range matches the
+                    // expectation.
+                    layout::Abi::Scalar(ref scalar) => {
+                        let size = scalar.value.size(self);
+                        let value = self.memory.read_scalar(ptr, ptr_align, size)?;
+                        self.validate_scalar(value, size, scalar, &path, layout.ty)?;
+                        if scalar.value == Primitive::Pointer {
+                            // ignore integer pointers, we can't reason about the final hardware
+                            if let Scalar::Ptr(ptr) = value {
+                                let alloc_kind = self.tcx.alloc_map.lock().get(ptr.alloc_id);
+                                if let Some(AllocType::Static(did)) = alloc_kind {
+                                    // statics from other crates are already checked
+                                    // extern statics should not be validated as they have no body
+                                    if !did.is_local() || self.tcx.is_foreign_item(did) {
+                                        return Ok(());
+                                    }
+                                }
+                                if let Some(tam) = layout.ty.builtin_deref(false) {
+                                    // we have not encountered this pointer+layout combination before
+                                    if seen.insert((ptr, tam.ty)) {
+                                        todo.push((ptr, tam.ty, format!("(*{})", path)))
+                                    }
+                                }
+                            }
+                        }
+                        Ok(())
+                    },
+                    _ => bug!("bad abi for FieldPlacement::Union(0): {:#?}", layout.abi),
                 }
             }
-
-            _ => (),
+            layout::FieldPlacement::Union(_) => {
+                // We can't check unions, their bits are allowed to be anything.
+                // The fields don't need to correspond to any bit pattern of the union's fields.
+                // See https://github.com/rust-lang/rust/issues/32836 for details
```

Comment:
Maybe link to a comment, because that's a huge thread? E.g. you could link to https://github.com/rust-lang/rust/issues/32836#issuecomment-406875389